File: portal/backend/internal/api/handlers/automation.go
package handlers

import (
	"encoding/json"
	"net/http"

	"github.com/go-chi/chi/v5"
	"kit.workmate/live-portal/internal/automation"
)

// AutomationHandler handles automation rule CRUD and test execution
type AutomationHandler struct {
	store  *automation.Store
	engine *automation.Engine
}

// NewAutomationHandler creates a new AutomationHandler
func NewAutomationHandler(store *automation.Store, engine *automation.Engine) *AutomationHandler {
	return &AutomationHandler{store: store, engine: engine}
}

// ListRules returns all automation rules
func (h *AutomationHandler) ListRules(w http.ResponseWriter, r *http.Request) {
	rules := h.store.List()
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(rules)
}

// CreateRule creates a new automation rule
func (h *AutomationHandler) CreateRule(w http.ResponseWriter, r *http.Request) {
	var rule automation.Rule
	if err := json.NewDecoder(r.Body).Decode(&rule); err != nil {
		http.Error(w, "Invalid request body", http.StatusBadRequest)
		return
	}
	if rule.Name == "" {
		http.Error(w, "Name is required", http.StatusBadRequest)
		return
	}
	if rule.Trigger.Type == "" {
		http.Error(w, "Trigger type is required", http.StatusBadRequest)
		return
	}
	if err := h.store.Add(rule); err != nil {
		http.Error(w, err.Error(), http.StatusConflict)
		return
	}
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(http.StatusCreated)
	json.NewEncoder(w).Encode(map[string]string{"status": "ok"})
}

// UpdateRule replaces an existing automation rule
func (h *AutomationHandler) UpdateRule(w http.ResponseWriter, r *http.Request) {
	name := chi.URLParam(r, "name")
	var rule automation.Rule
	if err := json.NewDecoder(r.Body).Decode(&rule); err != nil {
		http.Error(w, "Invalid request body", http.StatusBadRequest)
		return
	}
	if err := h.store.Update(name, rule); err != nil {
		http.Error(w, err.Error(), http.StatusNotFound)
		return
	}
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(map[string]string{"status": "ok"})
}

// DeleteRule removes an automation rule
func (h *AutomationHandler) DeleteRule(w http.ResponseWriter, r *http.Request) {
	name := chi.URLParam(r, "name")
	if err := h.store.Delete(name); err != nil {
		http.Error(w, err.Error(), http.StatusNotFound)
		return
	}
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(map[string]string{"status": "ok"})
}

// TestRule manually fires a rule by name with empty vars
func (h *AutomationHandler) TestRule(w http.ResponseWriter, r *http.Request) {
	name := chi.URLParam(r, "name")
	if err := h.engine.Test(name); err != nil {
		http.Error(w, err.Error(), http.StatusNotFound)
		return
	}
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(map[string]string{"status": "ok"})
}

File: portal/frontend/src/services/automation.ts
import { authFetch } from '@/lib/api'
import type { Rule } from '@/types/automation'

export const automationAPI = {
  async listRules(): Promise<Rule[]> {
    const res = await authFetch('/api/automation/rules')
    if (!res.ok) throw new Error('Failed to load rules')
    return res.json()
  },

  async createRule(rule: Rule): Promise<void> {
    const res = await authFetch('/api/automation/rules', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(rule),
    })
    if (!res.ok) {
      const text = await res.text()
      throw new Error(text || 'Failed to create rule')
    }
  },

  async updateRule(name: string, rule: Rule): Promise<void> {
    const res = await authFetch(`/api/automation/rules/${encodeURIComponent(name)}`, {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(rule),
    })
    if (!res.ok) {
      const text = await res.text()
      throw new Error(text || 'Failed to update rule')
    }
  },

  async deleteRule(name: string): Promise<void> {
    const res = await authFetch(`/api/automation/rules/${encodeURIComponent(name)}`, {
      method: 'DELETE',
    })
    if (!res.ok) throw new Error('Failed to delete rule')
  },

  async testRule(name: string): Promise<void> {
    const res = await authFetch(`/api/automation/rules/${encodeURIComponent(name)}/test`, {
      method: 'POST',
    })
    if (!res.ok) throw new Error('Failed to test rule')
  },
}
